### PR TITLE
Feature/42 participation rate filter

### PR DESCRIPTION
**Pull request by chrisbc**
closing #42 and closing #46 this PR includes:
 - new style filter package to standardise and improve filtering on various aspects of the solution datasets.
 - conditional vs non-conditional participaton
 - many docstrings and mkdocs improvements
 - test optimisations
 - dataframe models using the `panderas` library

### CODE REVIEW

File: test/participation/test_fault_system_solution_section_participation.py
Comment (chrisdicaprio) [inline test/participation/test_fault_system_solution_section_participation.py:1]: cleanup?

**Review by chrisdicaprio — APPROVED** [review]

File: solvis/filter/parent_fault_id_filter.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    Raises:
+        ValueError: If any member of `validate_names` argument is not valid.
+    """
+    unknown = set(validate_names).difference(set(solution.parent_fault_names))
+    if unknown:
+        raise ValueError(f"The solution {solution} does not contain the parent_fault_names: {unknown}.")
+    return set(validate_names)
+
+
+class FilterParentFaultIds:
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply solvis/filter/parent_fault_id_filter.py:73]: that one is deferred . see #51

File: solvis/filter/rupture_id_filter.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        Args:
+            solution: The solution instance to act on.
+            drop_zero_rates: Exclude ruptures with rupture_rate == 0 (default=True)
+        """
+        self._solution = solution
+        self._drop_zero_rates = drop_zero_rates
+        self._filter_subsection_ids = FilterSubsectionIds(solution)
+        self._filter_parent_fault_ids = FilterParentFaultIds(solution)
+
+    def all(self) -> ChainableSetBase:
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply solvis/filter/rupture_id_filter.py:58]: No, there is some sub-class specific stuff in there.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,26 @@
 # Changelog
 
+## [0.13.0] 2024-*
+
+## Changed
+ - drop_zero_rate differentiates between FSS and IS;
+ - faster testing (~3 times faster)
+ - documentation improvements
+ - abc427e fixing many deprecation warnings; updated geopandas, pandas, numpy, pyvista libs;
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply CHANGELOG.md:9]: removed

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,26 @@
...
+ - read_csv dtype configuration improvements;
+ - many old functions are deprecated/migrated to new filter package
+ - refactor dataframe dtypes
+ - standardise participation functions API and return columns;
+
+## Added
+ - new filter package providing classes for filtering solotions
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply CHANGELOG.md:16]: ta

File: solvis/geometry.py [hunk 5]
```diff
@@ -102,18 +102,68 @@ def create_surface(
...
+
+def fault_surface_projection(
+    trace: LineString,
+    dip_dir: float,
+    dip_deg: float,
+    upper_depth: float,
+    lower_depth: float,
+) -> Union[Polygon, LineString]:
+    """
+    Calculate a projection of the fault surface onto the geodetic sphere
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply solvis/geometry.py:117]: thanks

File: solvis/inversion_solution/dataframe_models.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    """A base schema for common dataframe columns.
+
+    This is not used directly, but it's fields are reused in other schema.
+
+    Attributes:
+     index: unique index on section_id.
+     section_id: the section id (OpenSHA:FaultID).
+     section_name: section name (OpenSHA:FaultName).
+     dip_angle: dip angle in degrees (OpenSHA:DipDeg).
+     rake: fault sectoin rake in degrees (OpenSHA:Rake).
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply solvis/inversion_solution/dataframe_models.py:61]: ta

File: solvis/inversion_solution/inversion_solution_operations.py [hunk 7]
```diff
@@ -106,74 +237,87 @@ def fault_sections_with_rupture_rates(self) -> gpd.GeoDataFrame:
...
 
     @property
-    def fault_sections_with_solution_slip_rates(self) -> gpd.GeoDataFrame:
-        """
-        Calculate and cache fault sections and their solution slip rates.
-        Solution slip rate combines input (avg slips) and solution (rupture rates).
+    def fault_sections_with_solution_slip_rates(self) -> 'DataFrame[FaultSectionWithSolutionSlipRate]':
+        """Calculate and cache fault sections and their solution slip rates.
+
+        NB: Solution slip rate combines input (avg slips) and solution (rupture rates).
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply solvis/inversion_solution/inversion_solution_operations.py:250]: ta

File: solvis/inversion_solution/inversion_solution_operations.py [hunk 8]
```diff
@@ -1,37 +1,163 @@
...
 
     def rupture_surface(self, rupture_id: int) -> gpd.GeoDataFrame:
         return SolutionSurfacesBuilder(self).rupture_surface(rupture_id)
 
+    def section_participation_rates(
+        self, subsection_ids: Optional[Iterable[int]] = None, rupture_ids: Optional[Iterable[int]] = None
+    ) -> 'DataFrame[SectionParticipationSchema]':
+        """Calculate the 'participation rate' for fault subsections.
+
+        Participation rate is the sum of rupture rate on each fault sections.
```
[the remark this answers is not in the document]
Comment (chrisbc) [reply solvis/inversion_solution/inversion_solution_operations.py:76]: thanks!

**Pull request merged**